src/core/util/options.js

```diff
@@ -183,6 +183,7 @@ strats.watch = function (parentVal: ?Object, childVal: ?Object): ?Object {
 strats.props =
 strats.methods =
 strats.inject =
+strats.provide =
```

Comment:
This is actually not enough because `provide` can also be a function that returns an object.

I think we should make `data` strategy reusable and use it for `provide`
Other plugins like vuefire will also benefit from this 🙂